README.md
# PostCSS Colors Only [![Build Status][ci-img]][ci]

[PostCSS] plugin to remove all rules except those which contain one or more colors.

This tool is useful if you are re-skinning a site with a new color scheme and need a starting point for a new stylesheet.

[PostCSS]: https://github.com/postcss/postcss
[ci-img]:  https://travis-ci.org/rsanchez/postcss-colors-only.svg
[ci]:      https://travis-ci.org/rsanchez/postcss-colors-only

```css
.foo {
  color: red;
  border: 1px solid #ab560f;
  font-size: 16px;
  background-image: linear-gradient(to-bottom, red, blue);
}

.bar {
  color: rgba(0, 128, 255, 0.5);
}

.baz {
  display: block;
}
```

```css
.foo {
  color: red;
  border-color: #ab560f;
  background-image: linear-gradient(to-bottom, red, blue);
}

.bar {
  color: rgba(0, 128, 255, 0.5);
}
```

This plugin will remove any CSS rules that do not contain a color (named, hex, rgb, rgba, hsl, or hsla) value. It looks at the following CSS properties for colors:

* `color`
* `background`
* `background-color`
* `background-image`
* `border`
* `border-top`
* `border-right`
* `border-bottom`
* `border-left`
* `border-color`
* `border-top-color`
* `border-right-color`
* `border-bottom-color`
* `border-left-color`
* `outline`
* `outline-color`
* `text-shadow`
* `box-shadow`

The following CSS properties will be transformed, leaving only the color part of the declaration:

* `background` → `background-color`
* `border` → `border-color`
* `border-top` → `border-top-color`
* `border-right` → `border-right-color`
* `border-bottom` → `border-bottom-color`
* `border-left` → `border-left-color`
* `outline` → `outline-color`

## Installation

[![NPM version](https://badge.fury.io/js/postcss-colors-only.svg)](https://www.npmjs.org/package/postcss-colors-only)

[Use npm](https://www.npmjs.org/doc/cli/npm-install.html).

```
npm install postcss-colors-only
```

## Usage

```javascript
var postcss = require('postcss');
var colorsOnly = require('postcss-colors-only');
var options = {
  withoutGrey: false, // set to true to remove rules that only have grey colors
  withoutMonochrome: false, // set to true to remove rules that only have grey, black, or white colors
};

postcss()
  .use(colorsOnly(options))
  .process('a { color: red; background: #FF0 url(foo.jpg); font-size: 12px; }')
  .css;
//=> 'a { color: red; background-color: #FF0; }'
```

## CLI

This plugin is also available as a standalone command line tool. See [css-colors-only](https://github.com/rsanchez/css-colors-only).

## License

Copyright (c) 2015 [Rob Sanchez](https://github.com/rsanchez)

Licensed under [the MIT License](./LICENSE).


CHANGELOG.md
## 0.0.1
* README updates.

## 0.0.0
* Initial release.


test/test.js
var postcss = require('postcss');
var expect  = require('chai').expect;

var plugin = require('../');

var test = function (input, output, opts, done) {
    postcss([ plugin(opts) ]).process(input).then(function (result) {
        expect(result.css).to.eql(output);
        expect(result.warnings()).to.be.empty;
        done();
    }).catch(function (error) {
        done(error);
    });
};

describe('postcss-colors-only', function () {

    it('should keep named color.', function (done) {
        test(
            'a { color: red; } p { display: block; }',
            'a { color: red; }',
            {},
            done
        );
    });

    it('should keep three-letter hex color.', function (done) {
        test(
            'a { color: #123; } p { display: block; }',
            'a { color: #123; }',
            {},
            done
        );
    });

    it('should keep six-letter hex color.', function (done) {
        test(
            'a { color: #123123; } p { display: block; }',
            'a { color: #123123; }',
            {},
            done
        );
    });

    it('should keep rgb color.', function (done) {
        test(
            'a { color: rgb(1, 2, 3); } p { display: block; }',
            'a { color: rgb(1, 2, 3); }',
            {},
            done
        );
    });

    it('should keep rgba color.', function (done) {
        test(
            'a { color: rgba(1, 2, 3, 0.5); } p { display: block; }',
            'a { color: rgba(1, 2, 3, 0.5); }',
            {},
            done
        );
    });

    it('should keep hsl color.', function (done) {
        test(
            'a { color: hsl(1, 2%, 3%); } p { display: block; }',
            'a { color: hsl(1, 2%, 3%); }',
            {},
            done
        );
    });

    it('should keep hsla color.', function (done) {
        test(
            'a { color: hsla(1, 2%, 3%, 0.5); } p { display: block; }',
            'a { color: hsla(1, 2%, 3%, 0.5); }',
            {},
            done
        );
    });

    it('should keep background-color.', function (done) {
        test(
            'a { background-color: red; } p { display: block; }',
            'a { background-color: red; }',
            {},
            done
        );
    });

    it('should keep border-color.', function (done) {
        test(
            'a { border-color: red; } p { display: block; }',
            'a { border-color: red; }',
            {},
            done
        );
    });

    it('should keep border-top-color.', function (done) {
        test(
            'a { border-top-color: red; } p { display: block; }',
            'a { border-top-color: red; }',
            {},
            done
        );
    });

    it('should keep border-right-color.', function (done) {
        test(
            'a { border-right-color: red; } p { display: block; }',
            'a { border-right-color: red; }',
            {},
            done
        );
    });

    it('should keep border-bottom-color.', function (done) {
        test(
            'a { border-bottom-color: red; } p { display: block; }',
            'a { border-bottom-color: red; }',
            {},
            done
        );
    });

    it('should keep border-left-color.', function (done) {
        test(
            'a { border-left-color: red; } p { display: block; }',
            'a { border-left-color: red; }',
            {},
            done
        );
    });

    it('should keep background-image.', function (done) {
        test(
            'a { background-image: linear-gradient(to bottom, red, blue); } ' +
                'p { display: block; }',
            'a { background-image: linear-gradient(to bottom, red, blue); }',
            {},
            done
        );
    });

    it('should keep outline-color.', function (done) {
        test(
            'a { outline-color: red; } p { display: block; }',
            'a { outline-color: red; }',
            {},
            done
        );
    });

    it('should keep text-shadow.', function (done) {
        test(
            'a { text-shadow: 1px 1px 2px black; } p { display: block; }',
            'a { text-shadow: 1px 1px 2px black; }',
            {},
            done
        );
    });

    it('should keep box-shadow.', function (done) {
        test(
            'a { box-shadow: 10px 5px 5px black; } p { display: block; }',
            'a { box-shadow: 10px 5px 5px black; }',
            {},
            done
        );
    });

    it('should change background to background-color.', function (done) {
        test(
            'a { background: red url(../foo.jpg) no-repeat center center; } ' +
                'p { display: block; }',
            'a { background-color: red; }',
            {},
            done
        );
    });

    it('should not change with multiple colors.', function (done) {
        test(
            'a { background: red url(../foo.jpg), blue url(../bar.jpg); } ' +
                'p { display: block; }',
            'a { background: red url(../foo.jpg), blue url(../bar.jpg); }',
            {},
            done
        );
    });

    it('should change outline to outline-color.', function (done) {
        test(
            'a { outline: 1px solid white; } p { display: block; }',
            'a { outline-color: white; }',
            {},
            done
        );
    });

    it('should change border to border-color.', function (done) {
        test(
            'a { border: 1px solid white; } p { display: block; }',
            'a { border-color: white; }',
            {},
            done
        );
    });

    it('should change border-top to border-top-color.', function (done) {
        test(
            'a { border-top: 1px solid white; } p { display: block; }',
            'a { border-top-color: white; }',
            {},
            done
        );
    });

    it('should change border-right to border-right-color.', function (done) {
        test(
            'a { border-right: 1px solid white; } p { display: block; }',
            'a { border-right-color: white; }',
            {},
            done
        );
    });

    it('should change border-bottom to border-bottom-color.', function (done) {
        test(
            'a { border-bottom: 1px solid white; } p { display: block; }',
            'a { border-bottom-color: white; }',
            {},
            done
        );
    });

    it('should change border-left to border-left-color.', function (done) {
        test(
            'a { border-left: 1px solid white; } p { display: block; }',
            'a { border-left-color: white; }',
            {},
            done
        );
    });

    it('should remove grey, but not black or white.', function (done) {
        test(
            'a { color: red; } p { color: grey; } h1 { color: black; }',
            'a { color: red; } h1 { color: black; }',
            { withoutGrey: true },
            done
        );
    });

    it('should remove grey', function (done) {
        test(
            'a { color: red; } p { color: grey; }',
            'a { color: red; }',
            { withoutGrey: true },
            done
        );
    });

    it('should remove gray', function (done) {
        test(
            'a { color: red; } p { color: gray; }',
            'a { color: red; }',
            { withoutGrey: true },
            done
        );
    });

    it('should remove lightgrey', function (done) {
        test(
            'a { color: red; } p { color: lightgrey; }',
            'a { color: red; }',
            { withoutGrey: true },
            done
        );
    });

    it('should remove lightgray', function (done) {
        test(
            'a { color: red; } p { color: lightgray; }',
            'a { color: red; }',
            { withoutGrey: true },
            done
        );
    });

    it('should remove dimgrey', function (done) {
        test(
            'a { color: red; } p { color: dimgrey; }',
            'a { color: red; }',
            { withoutGrey: true },
            done
        );
    });

    it('should remove dimgray', function (done) {
        test(
            'a { color: red; } p { color: dimgray; }',
            'a { color: red; }',
            { withoutGrey: true },
            done
        );
    });

    it('should remove darkgrey', function (done) {
        test(
            'a { color: red; } p { color: darkgrey; }',
            'a { color: red; }',
            { withoutGrey: true },
            done
        );
    });

    it('should remove darkgray', function (done) {
        test(
            'a { color: red; } p { color: darkgray; }',
            'a { color: red; }',
            { withoutGrey: true },
            done
        );
    });

    it('should remove three-letter hex grey', function (done) {
        test(
            'a { color: red; } p { color: #111; }',
            'a { color: red; }',
            { withoutGrey: true },
            done
        );
    });

    it('should remove six-letter hex grey', function (done) {
        test(
            'a { color: red; } p { color: #121212; border-color: #111111; }',
            'a { color: red; }',
            { withoutGrey: true },
            done
        );
    });

    it('should remove rgb grey', function (done) {
        test(
            'a { color: red; } p { color: rgb(1, 1, 1); }',
            'a { color: red; }',
            { withoutGrey: true },
            done
        );
    });

    it('should remove rgba grey', function (done) {
        test(
            'a { color: red; } p { color: rgba(1, 1, 1, 0.5); }',
            'a { color: red; }',
            { withoutGrey: true },
            done
        );
    });

    it('should remove hsl grey', function (done) {
        test(
            'a { color: red; } p { color: hsl(0, 0, 1%); }',
            'a { color: red; }',
            { withoutGrey: true },
            done
        );
    });

    it('should remove hsla grey', function (done) {
        test(
            'a { color: red; } p { color: hsla(0, 0, 1%, 0.5); }',
            'a { color: red; }',
            { withoutGrey: true },
            done
        );
    });

    it('should remove grey', function (done) {
        test(
            'a { color: red; } p { color: grey; }',
            'a { color: red; }',
            { withoutMonochrome: true },
            done
        );
    });

    it('should remove white', function (done) {
        test(
            'a { color: red; } p { color: white; }',
            'a { color: red; }',
            { withoutMonochrome: true },
            done
        );
    });

    it('should remove three-letter hex white', function (done) {
        test(
            'a { color: red; } p { color: #fFf; }',
            'a { color: red; }',
            { withoutMonochrome: true },
            done
        );
    });

    it('should remove six-letter hex white', function (done) {
        test(
            'a { color: red; } p { color: #fFfFfF; }',
            'a { color: red; }',
            { withoutMonochrome: true },
            done
        );
    });

    it('should remove rgb white', function (done) {
        test(
            'a { color: red; } p { color: rgba(255, 255, 255); }',
            'a { color: red; }',
            { withoutMonochrome: true },
            done
        );
    });

    it('should remove rgba white', function (done) {
        test(
            'a { color: red; } p { color: rgba(255, 255, 255, 0.5); }',
            'a { color: red; }',
            { withoutMonochrome: true },
            done
        );
    });

    it('should remove hsl white', function (done) {
        test(
            'a { color: red; } p { color: hsl(0, 0, 100%); }',
            'a { color: red; }',
            { withoutMonochrome: true },
            done
        );
    });

    it('should remove hsla white', function (done) {
        test(
            'a { color: red; } p { color: hsla(0, 0, 100%, 0.5); }',
            'a { color: red; }',
            { withoutMonochrome: true },
            done
        );
    });

    it('should remove black', function (done) {
        test(
            'a { color: red; } p { color: black; }',
            'a { color: red; }',
            { withoutMonochrome: true },
            done
        );
    });

    it('should remove three-letter hex black', function (done) {
        test(
            'a { color: red; } p { color: #000; }',
            'a { color: red; }',
            { withoutMonochrome: true },
            done
        );
    });

    it('should remove six-letter hex black', function (done) {
        test(
            'a { color: red; } p { color: #000000; }',
            'a { color: red; }',
            { withoutMonochrome: true },
            done
        );
    });

    it('should remove rgb black', function (done) {
        test(
            'a { color: red; } p { color: rgba(0, 0, 0); }',
            'a { color: red; }',
            { withoutMonochrome: true },
            done
        );
    });

    it('should remove rgba black', function (done) {
        test(
            'a { color: red; } p { color: rgba(0, 0, 0, 0.5); }',
            'a { color: red; }',
            { withoutMonochrome: true },
            done
        );
    });

    it('should remove hsl black', function (done) {
        test(
            'a { color: red; } p { color: hsl(0, 0, 0%); }',
            'a { color: red; }',
            { withoutMonochrome: true },
            done
        );
    });

    it('should remove hsla black', function (done) {
        test(
            'a { color: red; } p { color: hsla(0, 0, 0%, 0.5); }',
            'a { color: red; }',
            { withoutMonochrome: true },
            done
        );
    });

    it('should remove comment', function (done) {
        test(
            'a { color: red; } /* some comment */',
            'a { color: red; }',
            {},
            done
        );
    });

    it('should read nested @media rules', function (done) {
        test(
            'a { color: red; } @media (screen-only) { a { color: blue; } ' +
                'p { display: block; } }',
            'a { color: red; } @media (screen-only) { a { color: blue; } }',
            {},
            done
        );
    });

    it('should remove empty @media rule', function (done) {
        test(
            'a { color: red; } @media (screen-only) { p { display: block; } }',
            'a { color: red; }',
            {},
            done
        );
    });

    it('should remove non-media atrule', function (done) {
        test(
            'a { color: red; } @charset "UTF-8";',
            'a { color: red; }',
            {},
            done
        );
    });
});


index.js
'use strict';

var postcss = require('postcss');

module.exports = postcss.plugin('postcss-colors-only', function (options) {
    var cssColorList = require('css-color-list');
    var hexRegex = require('hex-color-regex');
    var rgbaRegex = require('rgba-regex');
    var rgbRegex = require('rgb-regex');
    var hslaRegex = require('hsla-regex');
    var hslRegex = require('hsl-regex');
    var util = require('util');

    var greyRegex =
    '^(' +
    '#([a-fA-F0-9]{1,2})\\2\\2' + '|' +
    '(dim|dark|light)?gr[ae]y' + '|' +
    'rgb\\((\\d+),\\s*\\4,\\s*\\4\\)' + '|' +
    'rgba\\((\\d+),\\s*\\5,\\s*\\5,\\s*\\d*(?:\\.\\d+)?\\)' + '|' +
    'hsl\\(\\s*0\\s*,\\s*0(\\.[0]+)?%\\s*,\\s*\\d*(?:\\.\\d+)?%\\)' + '|' +
    'hsla\\(0,\\s*0(\\.[0]+)?%,\\s*\\d*(?:\\.\\d+)?%,\\s*\\d*(?:\\.\\d+)?\\)' +
    ')$';

    var blackOrWhiteRegex =
    '^(' +
    '#([fF]{3}([fF]{3})?|000|000000)' + '|' +
    'black' + '|' +
    'white' + '|' +
    'rgb\\((0|255),\\s*\\4,\\s*\\4\\)' + '|' +
    'rgba\\((0|255),\\s*\\5,\\s*\\5,\\s*\\d*(?:\\.\\d+)?\\)' + '|' +
    'hsl\\(\\s*0\\s*,\\s*0(\\.[0]+)?%\\s*,\\s*(0|100)(\\.[0]+)?%\\)' + '|' +
    'hsla\\(0,\\s*0(\\.[0]+)?%,\\s*(0|100)(\\.[0]+)?%,\\s*\\d*(?:\\.\\d+)?\\)' +
    ')$';

    options = util._extend({
        withoutGrey:       false,
        withoutMonochrome: false
    }, options);

    function doesPropertyAllowColor(property) {
        var properties = [
            'color',
            'background',
            'background-color',
            'background-image',
            'border',
            'border-top',
            'border-right',
            'border-bottom',
            'border-left',
            'border-color',
            'border-top-color',
            'border-right-color',
            'border-bottom-color',
            'border-left-color',
            'outline',
            'outline-color',
            'text-shadow',
            'box-shadow'
        ];

        return properties.indexOf(property) > -1;
    }

    function transformProperty(declaration, colors) {
        var propertyTransformers = {
            background:      'background-color',
            border:          'border-color',
            'border-top':    'border-top-color',
            'border-right':  'border-right-color',
            'border-bottom': 'border-bottom-color',
            'border-left':   'border-left-color',
            outline:         'outline-color'
        };

        if (typeof propertyTransformers[declaration.prop] === 'undefined') {
            return;
        }

        if (colors.length === 1) {
            declaration.prop = propertyTransformers[declaration.prop];
            declaration.value = colors[0];
        }
    }

    function isColor(value) {
        var regex = new RegExp(
            '^(' +
            cssColorList().join('|') + '|' +
            rgbRegex().source + '|' +
            rgbaRegex().source + '|' +
            hslRegex().source + '|' +
            hslaRegex().source + '|' +
            hexRegex().source +
            ')$'
        );

        return regex.test(value);
    }

    function isColorBlackOrWhite(color) {
        return new RegExp(blackOrWhiteRegex).test(color);
    }

    function isColorGrey(color) {
        if (isColorBlackOrWhite(color)) {
            return false;
        }

        return new RegExp(greyRegex).test(color);
    }

    function isColorMonochrome(color) {
        return isColorBlackOrWhite(color) || isColorGrey(color);
    }

    function findColors(string) {
        var colors = [];
        var list = postcss.list.comma(string);
        var values = [];

        list.forEach(function (items) {
            postcss.list.space(items).forEach(function (item) {
                var regex = new RegExp(
                    '^' +
                    '(-webkit-|-moz-|-o-)?' +
                    '(repeating-)?' +
                    '(radial|linear)-gradient\\((.*?)\\)' +
                    '$'
                );

                var match = item.match(regex);

                if (match) {
                    values = values.concat(postcss.list.comma(match[4]));
                } else {
                    values.push(item);
                }
            });
        });

        values.forEach(function (value) {
            if (!isColor(value)) {
                return;
            }

            if (options.withoutMonochrome && isColorMonochrome(value)) {
                return;
            }

            if (options.withoutGrey && isColorGrey(value)) {
                return;
            }

            colors.push(value);
        });

        return colors;
    }

    function findColorsInDeclaration(declaration) {
        if (!doesPropertyAllowColor(declaration.prop)) {
            return [];
        }

        return findColors(declaration.value);
    }

    function processDeclaration(declaration) {
        var colors = findColorsInDeclaration(declaration);

        if (colors.length === 0) {
            declaration.remove();
        } else {
            transformProperty(declaration, colors);
        }
    }

    function processRule(rule) {
        rule.each(processDeclaration);

        if (rule.nodes.length === 0) {
            rule.remove();
        }
    }

    function processAtrule(atrule) {
        switch (atrule.name) {
            case 'media':
                atrule.each(processNode);

                if (atrule.nodes.length === 0) {
                    atrule.remove();
                }

                break;
            // @TODO deal with other types of atrules
            default:
                atrule.remove();
                break;
        }
    }

    function processNode(node) {
        switch (node.type) {
            case 'atrule':
                processAtrule(node);
                break;
            case 'rule':
                processRule(node);
                break;
            default:
                node.remove();
                break;
        }
    }

    return function (css) {
        css.each(processNode);
    };
});
